man/man1/cwb_setifar.1
.\" Man page generated from reStructuredText.
.
.TH "CWB_SETIFAR" "1" "Jan 14, 2019" "" "coherent WaveBurst"
.SH NAME
cwb_setifar \- cWB Documentation
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.nf

.fi
.sp
.nf

\fBUsage\fP

This command adds a new ifar (inverse false alarm rate) leaf to the
selected entries in the merged simulation root file.
For each selected entry a new leaf is created "ifar" which value is
obtained from the background FAR vs RHO file.
It is a text file which contains a corrispondence between far and rho (from background estimation)

\fBSyntax\fP

.fi
.sp
.INDENT 0.0
.INDENT 3.5
.INDENT 0.0
.IP \(bu 2
\fBcwb_setifar mlabel/list tsel farfile ilabel mode\fP
.IP \(bu 2
\fBmlabel/lists\fP same label used by cwb_merge, if the user writes
lists it list the available labels
.IP \(bu 2
.INDENT 2.0
.TP
\fBtsel\fP treen selections, only wave tree
must contains only leaf of waveburst tree (not the leaf run)
the syntax is the the same as the one used for the selection TTree
.UNINDENT
.IP \(bu 2
\fBilabel\fP label that will be added to the new merged files names (M1 \-> M1.S_ilabel)
.IP \(bu 2
.INDENT 2.0
.TP
\fBmode\fP inclusive/exclusive
inclusive : if a previous ifar is already associated to the entry
then it is updated if the new ifar is greatar
exclusive : the ifar is updated regardless what is is the old one
.UNINDENT
.UNINDENT
.UNINDENT
.UNINDENT
.nf

\fBAlternative Syntax\fP

.fi
.sp
.INDENT 0.0
.INDENT 3.5
.INDENT 0.0
.IP \(bu 2
\fBcwb_setifar mlabel/list options\fP
.IP \(bu 2
\fBmlabel/lists\fP same label used by cwb_merge
.IP \(bu 2
\fBoptions\fP options (options must be declared inside \(aq...\(aq) : see previous definitions
.UNINDENT
.INDENT 0.0
.INDENT 3.5
.INDENT 0.0
.IP \(bu 2
\fB\-\-label ilabel\fP default=""
.IP \(bu 2
\fB\-\-tsel tsel\fP default="" : tsel is the name or a logical
expression of names of TCut selections defined in
user_pparameters.C
.IP \(bu 2
\fB\-\-file file\fP default="" : file is the name of the far file
defined in user_pparameters.C, name can be an element of an array
(Ex: name[2])
.IP \(bu 2
\fB\-\-mode mode\fP default=""
.IP \(bu 2
\fB\-\-xtsel xtsel\fP default="" : the tsel are explicitly defined
(must be used as alternative to \-\-tsel)
.IP \(bu 2
\fB\-\-xfile xfile\fP default="" : the far file path is declared
explicitly (must be used as alternative to \-\-file)
.UNINDENT
.UNINDENT
.UNINDENT
.sp
\fBNOTE:\fP
.INDENT 0.0
.INDENT 3.5
label, tsel/xtsel, file/xfile, mode are mandatory
.UNINDENT
.UNINDENT
.sp
\fBNOTE:\fP
.INDENT 0.0
.INDENT 3.5
farfile is produced by the background report, and it is saved under the report data dir and its name is far_rho.txt
.UNINDENT
.UNINDENT
.sp
\fBNOTE:\fP
.INDENT 0.0
.INDENT 3.5
examples of allowed logical expressions are X&&Y, (X||Y), !X, ((X||Y)&&Z)
.UNINDENT
.UNINDENT
.UNINDENT
.UNINDENT
.nf

\fBExamples\fP

.fi
.sp
.INDENT 0.0
.INDENT 3.5
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
* cwb_setifar M1 \(aqnetcc[1]>0.8\(aq /home/waveburst/work_dir/report/postprod/far_rho.txt cc1_gt_0d8 exclusive
  \- add ifar leaf for all entries with ((netcc[1]>0.8)) in the wave file
  \- far is read from /home/frank/work_dir/report/postprod/far_rho.txt : is a two columns file "rho far"

  Alternative Syntax :

  cwb_setifar M1 \(aq\-\-xtsel netcc[1]>0.8  \-\-label cc1_gt_0d8 \e
              \-\-xfile /home/waveburst/work_dir/report/postprod/far_rho.txt \-\-mode exclusive\(aq

  or :

  cwb_setifar M1 \(aq\-\-tsel cc1_tsel  \-\-label cc1_gt_0d8 \-\-file farfile \-\-mode exclusive\(aq
  \- cc1_tsel is the name of TCut selection defined in user_pparameters.C
    Ex : TCut cc1_tsel("cc1_tsel","netcc[1]>0.8");
  \- farfile  is the name of the far file defined in user_pparameters.C
    Ex : TString farfile = "/home/waveburst/work_dir/report/postprod/far_rho.txt";

  or :

  cwb_setifar M1 \(aq\-\-tsel cc0_cut&&cc2_cut  \-\-label cc0_cc2_cut \-\-file farfile[2] \-\-mode exclusive\(aq
  \- cc0_cut and cc2_cut are the names of TCut selections defined in user_pparameters.C
    Ex : TCut cc0_cut("cc0_cut","netcc[0]>0.8");
         TCut cc2_cut("cc2_cut","netcc[2]>0.8");
  \- farfile[2] is the iname of the second element of the far file array defined in user_pparameters.C
    Ex : TString farfile[3];
         farfile[0] = "/home/waveburst/work_dir_0/report/postprod/far_rho.txt";
         farfile[1] = "/home/waveburst/work_dir_1/report/postprod/far_rho.txt";
         farfile[2] = "/home/waveburst/work_dir_2/report/postprod/far_rho.txt";
.ft P
.fi
.UNINDENT
.UNINDENT
.UNINDENT
.UNINDENT
.SH AUTHOR
Gabriele Vedovato
.SH COPYRIGHT
2018, cWB Team
.\" Generated by docutils manpage writer.
.


tools/cwb/macros/cwb_setifar.C
/*
# Copyright (C) 2019 Gabriele Vedovato
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/


// adds a new ifar (inverse alse alarm rate) leaf to the selected entries in the merged wave root file 
// for each selected entry a new leaf is created "ifar" which value is obtained from the CWB_SETFAR_FILE
// CWB_SETFAR_FILE is a text file which contains a corrispondence between rho and far (from background)

{
  int estat;
  Long_t id,size,flags,mt;
  char cmd[1024];

          cwb_merge_label      = TString(gSystem->Getenv("CWB_MERGE_LABEL"));
  TString cwb_setifar_tsel     = TString(gSystem->Getenv("CWB_SETIFAR_TSEL"));
  TString cwb_setifar_file     = TString(gSystem->Getenv("CWB_SETIFAR_FILE"));
  TString cwb_setifar_label    = TString(gSystem->Getenv("CWB_SETIFAR_LABEL"));
  TString cwb_setifar_mode     = TString(gSystem->Getenv("CWB_SETIFAR_MODE"));
  TString cwb_setifar_options  = TString(gSystem->Getenv("CWB_SETIFAR_OPTIONS"));

  TString mdir = merge_dir;

  if(cwb_setifar_options.Contains("--")) {
    TString option="";
    // get the cwb_setifar_tsel
    cwb_setifar_tsel = CWB::Toolbox::getParameter(cwb_setifar_options,"--tsel");
    if(cwb_setifar_tsel=="") {
      // look if the explicit tsel definition is defined
      cwb_setifar_tsel = CWB::Toolbox::getParameter(cwb_setifar_options,"--xtsel");
      if(cwb_setifar_tsel=="") {
        cout << "cwb_setifar.C : Error - (--tsel) or (--xtsel) is not defined !!!" << endl << endl;
        gSystem->Exit(1);
      }
    } else {
      // search global definition (defined in *.hh pp cuts file)
      // check if cwb_setifar_tsel is defined

      TString TSEL = cwb_setifar_tsel;

      cwb_setifar_tsel.ReplaceAll("&&"," ");
      cwb_setifar_tsel.ReplaceAll("||"," ");
      cwb_setifar_tsel.ReplaceAll("("," ");
      cwb_setifar_tsel.ReplaceAll(")"," ");
      cwb_setifar_tsel.ReplaceAll("!"," ");

      TObjArray* token = TString(cwb_setifar_tsel).Tokenize(TString(' '));
      for(int j=0;j<token->GetEntries();j++){

        TObjString* tok = (TObjString*)token->At(j);
        TString stok = tok->GetString();

         // check if tsel is defined
         TGlobal *global=(TGlobal*)gROOT->GetListOfGlobals()->FindObject(stok.Data());
         if(global==NULL) {
            cout << "cwb_setifar.C : Error - tsel \'" << stok << "\' -> wrong syntax or it is not defined " << endl;
            cout << "                must be included in the user_pparameters.C file" << endl << endl;
            gSystem->Exit(1);
         }
         TCut *tsel = (TCut*)global->GetAddress();
         if(TString(tsel->GetName())!=stok) {
            cout << "cwb_setifar.C : Error - tsel \'" << stok << "\' not correspond to global Tcut name" << endl;
            cout << "                must be included in the user_pparameters.C file" << endl << endl;
            gSystem->Exit(1);
         }

         TSEL.ReplaceAll(stok,TString::Format("(%s)",tsel->GetTitle()));
       }

       cwb_setifar_tsel = TSEL;    // extract TCut 
       cout << "cwb_setifar_tsel : " << cwb_setifar_tsel << endl;
    } 

    // get the cwb_setifar_file
    cwb_setifar_file = CWB::Toolbox::getParameter(cwb_setifar_options,"--file");
    if(cwb_setifar_file=="") {
      // look if the explicit file path definition is defined
      cwb_setifar_file = CWB::Toolbox::getParameter(cwb_setifar_options,"--xfile");
      if(cwb_setifar_file=="") {
        cout << "cwb_setifar.C : Error - (--file) or (--xfile) is not defined !!!" << endl << endl;
        gSystem->Exit(1);
      }
    } else {
      // search global definition far vs rho file
      // check if cwb_setifar_file is defined

      // check if cwb_setifar_file is an index of an array, ex: file_list[3] 
      int beg_bracket = cwb_setifar_file.Index("[");
      int end_bracket = cwb_setifar_file.Index("]");
      int len_bracket = end_bracket-beg_bracket+1;
      bool is_bracket=true;
      if(end_bracket!=cwb_setifar_file.Sizeof()-2) is_bracket=false;
      if(len_bracket<=0) is_bracket=false;
      TString sindx = cwb_setifar_file(beg_bracket+1,len_bracket-2);
      if(!sindx.IsDigit()) is_bracket=false;
      // extract array index
      int indx = is_bracket ? sindx.Atoi() : -1;
      if(indx>=0) cwb_setifar_file=cwb_setifar_file(0,beg_bracket);

      TGlobal *global=(TGlobal*)gROOT->GetListOfGlobals()->FindObject(cwb_setifar_file.Data());
      if(global==NULL) {
         cout << "cwb_setifar.C : Error - setifar_file \'" << cwb_setifar_file << "\' -> wrong syntax or is not defined " << endl;
         cout << "                must be included in the user_pparameters.C file" << endl << endl;
         gSystem->Exit(1);
      }
      int adim = global->GetArrayDim();
      if(adim==0) { 	// string
        cwb_setifar_file = *(TString*)global->GetAddress();
      }
      if(adim==1) { 	// array of strings dim=1
        int asize = global->GetMaxIndex(0); 
        if(indx<0 || indx>=asize) {
          cout << "cwb_setifar.C : Error - setifar_file \'" << cwb_setifar_file << "\' input array index not declared or not allowed, max=" << asize-1 << endl;
          cout << "                check user_pparameters.C file" << endl << endl;
          gSystem->Exit(1);
        }
        TString* afile = (TString*)global->GetAddress();
        cwb_setifar_file = afile[indx]; 
      }
      if(adim>1) {	// array of strings dim>1 (not allowed)
         cout << "cwb_setifar.C : Error - setifar_file \'" << cwb_setifar_file << "\' array dimension > 1 not allowed " << endl;
         cout << "                check user_pparameters.C file" << endl << endl;
         gSystem->Exit(1);
      } 
      cout << "cwb_setifar_file : " << cwb_setifar_file << endl;
    }

    // get the cwb_setifar_mode
    cwb_setifar_mode = CWB::Toolbox::getParameter(cwb_setifar_options,"--mode");
    if(cwb_setifar_mode=="") {
      cout << "cwb_setifar.C : Error - --mode is not defined !!!" << endl;
      gSystem->Exit(1);
    }

    // get the cwb_setifar_label
    cwb_setifar_label = CWB::Toolbox::getParameter(cwb_setifar_options,"--label");
    if(cwb_setifar_label=="") {
      cout << "cwb_setifar.C : Error - --label is not defined !!!" << endl;
      gSystem->Exit(1);
    }
  }

  // set inclusive mode 
  bool inclusive=true;
  cwb_setifar_mode.ToUpper();
  if(cwb_setifar_mode.BeginsWith("I")) inclusive=true;  
  if(cwb_setifar_mode.BeginsWith("E")) inclusive=false;  

  // check if cwb_ifar_label is defined
  if(cwb_setifar_label=="") {
    cout << "cwb_setifar.C : Error - ifar label not defined" << endl; 
    gSystem->Exit(1);
  }
  // check if cwb_setifar_label contains '.'
  if(cwb_setifar_label.Contains(".")) {
    cout << "cwb_setifar.C : Error - cwb_setifar_label " << cwb_setifar_label 
         << " can not contains '.'" << endl << endl;
    gSystem->Exit(1);
  }
  if(cwb_setifar_tsel=="") cwb_wcuts_tree="run>0";

  // create input wave root wave file name
  char iwfname[1024];  
  sprintf(iwfname,"wave_%s.%s.root",data_label,cwb_merge_label.Data());

  // create output wave root cuts file name
  TString owfname = mdir+"/"+iwfname;
  owfname.ReplaceAll(".root",TString(".S_")+cwb_setifar_label+".root");

  // apply setifar to the selected entries in the wave file
  int nsel = CWB::Toolbox::setIFAR(iwfname,mdir,mdir,"waveburst",cwb_setifar_tsel,
                                   cwb_setifar_file,pp_irho,cwb_setifar_label,inclusive);
  if(nsel<=0) {
    cout << "cwb_setifar.C : Warninig - Number of selected waveburst entries = " << nsel << endl << endl;
    //gSystem->Exit(1);
  }
  cout << "cwb_setifar.C : Number of selected waveburst entries = " << nsel << endl;

  // if cwb_setifar_label=same the original wave root file is overwrite
  if(cwb_setifar_label=="same") {
    sprintf(cmd,"/bin/mv %s %s/%s", owfname.Data(), mdir.Data(), iwfname);
    cout << cmd << endl;
    gSystem->Exec(cmd);
    exit(0);
  }

  // create a merge*.lst file name & run selection cuts
  vector<int> jobList;
  char ilstfname[1024];  
  sprintf(ilstfname,"merge_%s.%s.lst",data_label,cwb_merge_label.Data());
  TString olstfname = owfname;
  olstfname.ReplaceAll("wave_","merge_");
  olstfname.ReplaceAll(".root",".lst");
  olstfname.Remove(0,olstfname.Last('/')+1);	// strip path
  cout << olstfname << endl;
  estat = gSystem->GetPathInfo(mdir+"/"+ilstfname,&id,&size,&flags,&mt);
  if (estat==0) {
    sprintf(cmd,"cd %s;ln -sf %s %s",mdir.Data(),ilstfname,olstfname.Data());
    cout << cmd << endl;
    gSystem->Exec(cmd);
  }

  // for simulation!=0 create a symbolic link to mdc*.root file name
  if(simulation) {
    char ilfname[1024];
    sprintf(ilfname,"mdc_%s.%s.root",data_label,cwb_merge_label.Data());
    TString olfname = owfname;
    olfname.ReplaceAll("wave_","mdc_");
    olfname.Remove(0,olfname.Last('/')+1);	// strip path
    cout << olfname << endl;
    estat = gSystem->GetPathInfo(mdir+"/"+ilfname,&id,&size,&flags,&mt);
    if(estat==0) {
      sprintf(cmd,"cd %s;ln -sf %s %s",mdir.Data(),ilfname,olfname.Data());
      cout << cmd << endl;
      gSystem->Exec(cmd);
    }
  }

  // for simulation=0 create a symbolic link to live*.root file name
  if(!simulation) {
    char ilfname[1024];
    sprintf(ilfname,"live_%s.%s.root",data_label,cwb_merge_label.Data());
    TString olfname = owfname;
    olfname.ReplaceAll("wave_","live_");
    olfname.Remove(0,olfname.Last('/')+1);	// strip path
    cout << olfname << endl;
    estat = gSystem->GetPathInfo(mdir+"/"+ilfname,&id,&size,&flags,&mt);
    if(estat==0) {
      sprintf(cmd,"cd %s;ln -sf %s %s",mdir.Data(),ilfname,olfname.Data());
      cout << cmd << endl;
      gSystem->Exec(cmd);
    }
  }

  exit(0);
}
